### PR TITLE
Lending Strategy

### PR DESCRIPTION
**Pull request by jmonteer**

### CODE REVIEW

File: contracts/Strategy.sol [hunk 1]
```diff
@@ -20,44 +22,119 @@ import {
...
-        // TODO: calculate how much would it cost to close NOW
-        // TODO: check value of profits (and decide if we want to include it here)
-
-        return want.balanceOf(address(this));
+        // To estimate the assets under management of the strategy we add the want balance already 
+        // in the contract and the current valuation of the non-matured positions (including the cost of)
+        // closing the position early
+        // This function is supposed to be called after _checkPositionsAndWithdraw() so the matured positions 
+        // are supposed to already be included in the contract's want balance
+        return balanceOfWant()
```
Comment (fp-crypto) [inline contracts/Strategy.sol:373]: Should this include ETH balance when the want is WETH?

File: contracts/Strategy.sol [hunk 2]
```diff
@@ -67,123 +144,343 @@ contract Strategy is BaseStrategy {
...
+            _profit = 0;
+        }
+        uint256 amountRequired = _debtOutstanding.add(_profit);
+        if(amountRequired > wantBalance) {
+            // we need to free funds
+            // NOTE: liquidatePosition will try to use balanceOfWant first
+            // liquidatePosition will realise Losses if required !! (which cannot be equal to unrealised losses if
+            // we are not withdrawing 100% of position)
+            (uint256 amountAvailable, uint256 realisedLoss) = liquidatePosition(amountRequired);
+            _loss = realisedLoss;
```
Comment (fp-crypto) [inline contracts/Strategy.sol]: Under what conditions would you expect a loss?

File: contracts/Strategy.sol [hunk 3]
```diff
@@ -67,123 +144,343 @@ contract Strategy is BaseStrategy {
...
+
+        return result;
     }
+    
+    /*
+     * @notice
+     *  Internal function to assess the unrealised P&L of the Notional's positions
+     * @return uint256 result, the encoded trade ready to be used in Notional's 'BatchTradeAction'
+     */
+    function getUnrealisedPL() internal returns (uint256 _unrealisedProfit, uint256 _unrealisedLoss) {
```
Comment (fp-crypto) [inline contracts/Strategy.sol]: This can be a view

File: contracts/Strategy.sol [hunk 4]
```diff
@@ -67,123 +144,343 @@ contract Strategy is BaseStrategy {
...
+        uint256 amountToLiquidate = _amountNeeded.sub(wantBalance);
+        
+        // Losses are realised IFF we withdraw from the position, as they will come from breaking our "promise"
+        // of lending at a certain %
+        // The strategy will only realise losses proportional to the amount we are liquidating
+        uint256 totalDebt = vault.strategies(address(this)).totalDebt;
+        uint256 lossesToBeRealised = unrealisedLosses.mul(amountToLiquidate).div(totalDebt.sub(wantBalance));
+        
+        // Due to how Notional works, we need to substract losses from the amount to liquidate
+        // If we don't do this and withdraw a small enough % of position, we will not incur in losses,
```
Comment (fp-crypto) [inline contracts/Strategy.sol:660]: Has any thought been given to having a toggle that indicates whether losses should be taken when liquidating versus avoiding lossess and only liquidating funds freeable without loss?

File: contracts/Strategy.sol [hunk 5]
```diff
@@ -196,25 +493,203 @@ contract Strategy is BaseStrategy {
...
+        }
+    }
+
+    /*
+     * @notice
+     *  Internal function to control access
+     */
+    function _onlyGovernance() internal {
+        require(msg.sender == governance());
+    }
```
Comment (fp-crypto) [inline contracts/Strategy.sol]: Is this being used over the modifier to save bytecode size?

File: contracts/Strategy.sol [hunk 6]
```diff
@@ -20,44 +22,119 @@ import {
...
+        (Token memory assetToken, Token memory underlying) = _nProxy.getCurrency(_currencyID);
+        DECIMALS_DIFFERENCE = uint256(underlying.decimals).mul(MAX_BPS).div(uint256(assetToken.decimals));
+        
+        // Check whether the currency is set up right
+        if (_currencyID == 1) {
+            require(address(0) == underlying.tokenAddress); 
+        } else {
+            require(address(want) == underlying.tokenAddress);
+        }
     }
```
Comment (fp-crypto) [inline contracts/Strategy.sol]: Is the choice to use a constructor over an initializer pattern intentional? If you use the initializer pattern, adding cloning is practically free.

File: contracts/Strategy.sol (hunk 2 above)
Comment (fp-crypto) [reply contracts/Strategy.sol]: Hmm, I see below. See comment https://github.com/jmonteer/notional-lending-strategy/pull/1#discussion_r788263389

File: contracts/Strategy.sol (hunk 3 above)
Comment (16slim) [reply contracts/Strategy.sol]: Yes!

File: contracts/Strategy.sol (hunk 4 above)
Comment (16slim) [reply contracts/Strategy.sol:660]: Yes, the only time funds are freeable without a loss is when maturity is reached, I included a toggle in liquidatePosition that defines whether to realize losses or not. With this, the only option to realize losses is to really know what you're doing + explicitly wanting to do so as it needs to be changed to true

File: contracts/Strategy.sol (hunk 2 above)
Comment (16slim) [reply contracts/Strategy.sol]: Conversation below

File: contracts/Strategy.sol (hunk 5 above)
Comment (16slim) [reply contracts/Strategy.sol]: There were no modifiers implemented in BaseStrategy last week but I see it's been fixed by @jmonteer now, removed it and used the inherited modifier

File: contracts/Strategy.sol (hunk 1 above)
Comment (16slim) [reply contracts/Strategy.sol:373]: Shouldn't be needed as all the ETH interactions with Notional are wrapped with `weth.withdraw` or `weth.deposit` so the strategy never has an ETH balance. Just in case the strategy also implements a function to sweep any potential ETH to governance (function `sendETHToGovernance`)

File: contracts/Strategy.sol (hunk 4 above)
Comment (16slim) [reply contracts/Strategy.sol:660]: Toggle is included as a state variable called `toggleRealizeLosses`, only changeable by governance

File: contracts/Strategy.sol (hunk 6 above)
Comment (16slim) [reply contracts/Strategy.sol]: Following your repos and @jmonteer 's examples I included the initializer pattern to be able to clone the contract and deploy the strategy for different tokens/ vaults!

File: contracts/Strategy.sol [hunk 7]
```diff
@@ -67,123 +143,343 @@ contract Strategy is BaseStrategy {
...
+        }
 
+        // Use the market index with the shortest maturity
+        uint256 minMarketIndex = _getMinimumMarketIndex();
+        // NOTE: May revert if the availableWantBalance is too high and interest rates get to < 0
+        int256 fCashAmountToTrade = nProxy.getfCashAmountGivenCashAmount(
+            _currencyID, 
+            -int88(availableWantBalance.mul(MAX_BPS).div(DECIMALS_DIFFERENCE)), 
+            minMarketIndex, 
+            block.timestamp + 5
```
[the remark this answers is not in the document]
Comment (16slim) [reply contracts/Strategy.sol]: The function performs an estimation and may have a small % error and as we are depositing exactly the amount, if it overestimates even by 1 unit it reverts, as per jeff recommendation included a scaling factor to ensure it does not revert

File: contracts/Strategy.sol [hunk 8]
```diff
@@ -20,44 +22,257 @@ import {
...
         // Add your own name here, suggestion e.g. "StrategyCreamYFI"
         return "StrategyNotionalLending";
     }
+    
+    /*
+     * @notice
+     *  Getter function for the current invested maturity
+     * @return uint256, current maturity we are invested in
+     */
+    function getMaturity() external view onlyGovernance returns(uint256) {
```
Comment (jmonteer) [inline contracts/Strategy.sol]: why is this onlyGovernance?

File: contracts/Strategy.sol [hunk 9]
```diff
@@ -20,44 +22,257 @@ import {
...
 
-        // OPTIONAL:
-        // TODO: calculate how much would it cost to close NOW
-        // TODO: check value of profits (and decide if we want to include it here)
+    /*
+     * @notice
+     *  Setter function for the toggle defining whether to realize losses or not
+     * @param _newToggle, new booelan value for the toggle
+     */
+    function setToggleRealizeLosses(bool _newToggle) external onlyGovernance {
```
Comment (jmonteer) [inline contracts/Strategy.sol]: move this to onlyEmergencyAuthorized

governance is too slow for this to be used in an emergency scenario

File: contracts/Strategy.sol [hunk 10]
```diff
@@ -20,44 +22,257 @@ import {
...
+        return minTimeToMaturity;
+    }
 
-        return want.balanceOf(address(this));
+    /*
+     * @notice
+     *  Setter function for the minimum time to maturity to invest into, accesible only to governance
+     * @param _newTime, new minimum time to maturity to invest into
+     */
+    function setMinTimeToMaturity(uint256 _newTime) external onlyGovernance {
```
Comment (jmonteer) [inline contracts/Strategy.sol]: change this to onlyEmergencyAuthorized

**Review by jmonteer — COMMENTED** [review]
regarding roles

File: contracts/Strategy.sol (hunk 10 above)
Comment (16slim) [reply contracts/Strategy.sol]: Done

File: contracts/Strategy.sol (hunk 9 above)
Comment (16slim) [reply contracts/Strategy.sol]: Done

File: contracts/Strategy.sol (hunk 8 above)
Comment (16slim) [reply contracts/Strategy.sol]: Opened it to external without access restrictions

File: contracts/Strategy.sol [hunk 11]
```diff
@@ -67,123 +365,483 @@ contract Strategy is BaseStrategy {
...
+        if(amountRequired > wantBalance) {
+            // we need to free funds
+            // NOTE: liquidatePosition will try to use balanceOfWant first
+            // liquidatePosition will realise Losses if required !! (which cannot be equal to unrealised losses if
+            // we are not withdrawing 100% of position)
+            uint256 amountAvailable = wantBalance;
+            uint256 realisedLoss = 0;
 
-        // TODO: report loss, profit, debtPayment
+            // If the toggle to realize losses is off, do not close any position
```
Comment (jmonteer) [inline contracts/Strategy.sol:417]: add a comment here to explain why toggleRealizeProfits is needed here?

File: contracts/Strategy.sol [hunk 12]
```diff
@@ -67,123 +365,483 @@ contract Strategy is BaseStrategy {
...
+            if(toggleRealizeLosses || toggleRealizeProfits) {
+                (amountAvailable, realisedLoss) = liquidatePosition(amountRequired);
+            }
+            _loss = realisedLoss;
```
Comment (jmonteer) [inline contracts/Strategy.sol]: does it make sense to move this `_loss = realisedLoss` into the if clause?

Comment (jmonteer) [reply contracts/Strategy.sol]: (minor gas savings)

File: contracts/Strategy.sol [hunk 13]
```diff
@@ -196,25 +854,251 @@ contract Strategy is BaseStrategy {
...
+    /*
+     * @notice
+     *  Internal function Closing a current non-mature position to re-invest the amount into a new 
+     * higher maturity market
+     * @param _currentMaturity, current maturity the strategy is invested in
+     * @return uint256, liberated amount, now existing in want balance to add up to the availableWantBalance
+     * to trade into in adjustPosition()
+     */
+    function _rollOverTrade(uint256 _currentMaturity) internal returns(uint256) {
+        uint256 prevBalance = balanceOfWant();
```
Comment (jmonteer) [inline contracts/Strategy.sol]: minor gas savings:

you can remove this and the return statement from here. 

As the only time this is called is in adjustPosition, inside an if, you can just update 
`availableWantBalance = balanceOfWant()` after calling `_rollOverTrade(_maturity)`

that way you save 1 balanceOfWant() call

File: contracts/Strategy.sol [hunk 14]
```diff
@@ -67,123 +365,483 @@ contract Strategy is BaseStrategy {
...
-        // TODO: research how to pass this as calldata
-        BalanceActionsWithTrades[] memory actions = new BalanceActionsWithTrades[](1);
+        if (_currencyID == ONE) {
+            // Only necessary for wETH/ ETH pair
+            weth.withdraw(availableWantBalance);
+        } else {
+            want.approve(address(nProxy), availableWantBalance);
+        }
+        // Amount to trade is the available want balance, changed to 8 decimals and
+        // scaled down by FCASH_SCALING to ensure it does not revert
```
Comment (jmonteer) [inline contracts/Strategy.sol:489]: does this mean that some want will be held in strategy? 

we want to input 100 DAI but we calculate the amount of fcash that 99.95 DAI would be worth. is this right? 
so end of the day we are keeping 0.05 DAI in strategy.

(numbers are dummy)

File: contracts/Strategy.sol [hunk 15]
```diff
@@ -196,25 +854,251 @@ contract Strategy is BaseStrategy {
...
+    /*
+     * @notice
+     *  Public function used by the keeper to assess whether a harvest is necessary or not, 
+     * returns true only if there is a position to settle
+     * @param callCostInWei, call cost estimation performed by the keeper
+     * @return bool, true when the strategy has a mature position
+     */
+    function harvestTrigger(uint256 callCostInWei) public view override returns (bool) {
+        // Check is there is credit available for the strategy to invest
+        if (vault.creditAvailable() > 0) {
```
Comment (jmonteer) [inline contracts/Strategy.sol]: add a minimum amount here ! maybe as a percentage of totalAssets?

otherwise it will return true very easily

File: scripts/lendETH.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
+from pathlib import Path
```
Comment (jmonteer) [inline scripts/lendETH.py]: remove this file before merging

File: tests/test_migration.py [hunk 17]
```diff
@@ -17,28 +14,211 @@ def test_migration(
...
+
+    chain.sleep(3600 * 6)
+    chain.mine(1)
 
-    # TODO: add other tokens balance
-    pre_want_balance = token.balanceOf(strategy)
 
+    assert token.balanceOf(vault) > amount_invested
+
+def test_force_migration(
```
Comment (jmonteer) [inline tests/test_migration.py]: move to manual operation

File: tests/test_migration.py [hunk 18]
```diff
@@ -17,28 +14,211 @@ def test_migration(
...
+    strategy.setForceMigration(True, {"from": gov})
+    vault.migrateStrategy(strategy, new_strategy, {"from": gov})
+
+    want_balance_end = token.balanceOf(new_strategy)
+
+    assert want_balance_end > amount_invested
+
+    chain.mine(1, timedelta=3_600 * 6)
+
+def test_force_liquidations(
```
Comment (jmonteer) [inline tests/test_migration.py]: move to test_manual_operations.py

File: tests/test_migration.py [hunk 19]
```diff
@@ -17,28 +14,211 @@ def test_migration(
...
+
+    chain.mine(1, timedelta=2*86_400)
+    tx = strategy.liquidatefCashAmount(min_market_index, (fCash_balance - fCash_to_close), {"from":gov})
+    new_account = n_proxy_views.getAccount(strategy)
+
+    assert new_account[2] == []
+
+    chain.mine(1, timedelta=6 * 3_600)
+
+def test_emergency_exit(
```
Comment (jmonteer) [inline tests/test_migration.py]: move to test_manual_operations.py

**Review by jmonteer — COMMENTED** [review]
looks very good!! good job

File: contracts/Strategy.sol (hunk 11 above)
Comment (16slim) [reply contracts/Strategy.sol:417]: Yep, done

File: contracts/Strategy.sol (hunk 12 above)
Comment (16slim) [reply contracts/Strategy.sol]: I'm not sure because if we do not enter the if clause, _loss need to be set to 0 (coming from realisedLoss in line just before if clause)

File: contracts/Strategy.sol (hunk 14 above)
Comment (16slim) [reply contracts/Strategy.sol:489]: Yes, that's correct

File: tests/test_migration.py (hunk 17 above)
Comment (16slim) [reply tests/test_migration.py]: Yep, makes sense

File: scripts/lendETH.py (hunk 16 above)
Comment (16slim) [reply scripts/lendETH.py]: Yep, helper script from before

File: contracts/Strategy.sol (hunk 12 above)
Comment (16slim) [reply contracts/Strategy.sol]: I removed realisedLoss and replace the liquidatePosition line by:
(amountAvailable, _loss) = liquidatePosition(amountRequired);

File: contracts/Strategy.sol (hunk 13 above)
Comment (16slim) [reply contracts/Strategy.sol]: Correct, changed it. Thanks

File: contracts/Strategy.sol (hunk 12 above)
Comment (jmonteer) [reply contracts/Strategy.sol]: learning here: all non-initialized variables are initialized with the "false" value (aka 0 in uints)

File: contracts/Strategy.sol (hunk 15 above)
Comment (16slim) [reply contracts/Strategy.sol]: Good point, I added a MIN_AMOUNT_HARVEST state variable (setter for onlyVaultManagers) that is set during deployment to handle this

File: contracts/Strategy.sol (hunk 12 above)
Comment (16slim) [reply contracts/Strategy.sol]: Correct
